### PR TITLE
Tag enabled cost 1226

### PR DESCRIPTION
**Pull request by ddonahue007**
closes: https://issues.redhat.com/browse/COST-1226
closes: https://issues.redhat.com/browse/COST-1694
closes: https://issues.redhat.com/browse/COST-1871

Add Dual List Tree support to Applications Settings for Koku

https://user-images.githubusercontent.com/57504257/118814099-c88c6900-b87d-11eb-80be-2f05dc34366d.mov

Refactored for currency settings as well...

Screen Shot 2021-09-29 at 7 55 45 AM

Screen Shot 2021-09-29 at 7 56 45 AM

## Testing

[Smoke Tests](https://ci.ext.devshift.net/job/project-koku-koku-pr-check/1609/console)

```
23:17:14 =========================== short test summary info ============================
23:17:14 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_cost_reports.py::test_api_ocp_cost_filtered_total_match_group_by_total
23:17:14 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test_user.py::test_api_settings_ocp_tag_enablement
23:17:14 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test_user.py::test_api_settings_aws_tag_enablement
23:17:14 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test_user.py::test_api_settings_azure_tag_enablement
23:17:14 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test_user.py::test_api_settings_gcp_tag_enablement
23:17:14 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test_user.py::test_api_settings_negative_tag_enablement
23:17:14 = 6 failed, 5083 passed, 172 skipped, 860 deselected, 5 warnings in 3308.31s (0:55:08) =
```

### CODE REVIEW

**Comment by ddonahue007** [comment]
There are several PF issues opened against DualList Selector component.

**Comment by boaz0** [comment]
@ddonahue007 you need to update the following unit tests:
* test_post_settings_ocp_tag_disabled (api.settings.test.tests_views.SettingsViewTest)
Test setting OCP tags get disabled.

* test_post_settings_ocp_tag_enabled (api.settings.test.tests_views.SettingsViewTest)
Test setting OCP tags as enabled.

File: koku/api/settings/tag_management.py
Comment (myersCody) [inline koku/api/settings/tag_management.py:2]: Just as a heads up @ddonahue007, we switched our project license a while back from GPL to Apache. So we actually need the license header from before: 
```
#
# Copyright 2021 Red Hat Inc.
# SPDX-License-Identifier: Apache-2.0
#
```

Comment (ddonahue007) [reply koku/api/settings/tag_management.py]: Thanks

**Comment by ddonahue007** [comment]
This PR is still blocked by Pattern fly issues on the Dual List Selector component
https://github.com/patternfly/patternfly-react/issues?q=is%3Aissue+is%3Aopen+dual+in%3Atitle+

File: koku/api/settings/test/tests_views.py
Comment (dlabrecq) [inline koku/api/settings/test/tests_views.py:114]: Was the test code remove because it no longer applies?
Is there another test we can replace this with?

File: koku/api/settings/settings.py
Comment (dlabrecq) [inline koku/api/settings/settings.py:190]: It looks odd to have the currency select span the width of the page. 

Check with M. Coker, but I believe the Select is intended to fill the width of its container, but not the entire width of the card?

FYI, the Select components used in our details pages don't expand the width of the page. That's because the width is restricted by the toolbar's "item" container.

In that case, PatternFly's `--pf-c-toolbar__item--Width--base` variable is setting a `min-width` of 232px.

**Review by dlabrecq — COMMENTED** [review]
I'm not familiar enough with the DDF code to provide a thorough review, but will defer to others here. These two new features probably should be split into separate PRs? If there are more issues with the dual-list component, there is a possibility we may need to back out both changes.

File: koku/api/settings/settings.py
Comment (ddonahue007) [reply koku/api/settings/settings.py:190]: This is going to be handled in another PR, I am waiting on DDF to include the dropdown component. The issue here is that the component being used is the PF Select component. There are tow things i am waiting on from DDF project owners

1. There is no obvious way to over ride stying, and if there is can they give me an example.
2. Can the PF dropdown component be added, and is there docs on porting. (If they can not get it in soon rather then later.

I agree it looks odd but at this point this is what we have until either of the two above are taking care of. And it would be nice to get this in.

File: koku/api/settings/test/tests_views.py
Comment (ddonahue007) [reply koku/api/settings/test/tests_views.py:114]: The test code that was in place no longer applies and needs to be fully reworked. The settings page is now more then just the enable tabs object. I have added a @TODO for this work

Comment (dlabrecq) [reply koku/api/settings/test/tests_views.py:114]: Can you please create a JIRA issue?

File: koku/api/settings/settings.py
Comment (ddonahue007) [reply koku/api/settings/settings.py:191]: I have gotten the style working here...

File: koku/api/settings/test/tests_views.py
Comment (ddonahue007) [reply koku/api/settings/test/tests_views.py:114]: Added additional unit test coverage

File: koku/api/settings/settings.py
Comment (dlabrecq) [inline koku/api/settings/settings.py:277]: Is the log output to the browser console? If so, it's not desirable for production.

File: koku/api/settings/utils.py
Comment (dlabrecq) [inline koku/api/settings/utils.py:135]: What is the user experience if the code encounters a bad currency code? For example, is an alert shown?

File: koku/api/settings/settings.py
Comment (dlabrecq) [inline koku/api/settings/settings.py:286]: I'm trying to understand why setting the currency would generate an exception, considering it's not a user input?

**Comment by dlabrecq** [comment]
Looks like there is an IndexError with `get_tag_settings`

File: koku/api/settings/settings.py
Comment (adberglund) [reply koku/api/settings/settings.py:277]: Would only be logged to the pod logs in OpenShift

File: koku/api/settings/utils.py
Comment (ddonahue007) [reply koku/api/settings/utils.py:135]: No this would never be triggers by the UI at all. The only way this could ever be triggered is if you called the API directly with a currency that is not supported.

File: koku/api/settings/settings.py
Comment (ddonahue007) [reply koku/api/settings/settings.py:286]: It is not anything the UI will ever hit, it is just good practice when writing any API to make sure you cover yourself with exception handling.

File: koku/currency/__init__.py
Comment (maskarb) [inline koku/currency/__init__.py:1]: praise: I like that this `currency` folder was removed!

**Comment by ddonahue007** [comment]
Pay no attention to the test failures, I have already worked with QE on updating the smoke-tests.

File: koku/api/settings/utils.py
Comment (maskarb) [inline koku/api/settings/utils.py:111]: chore: let's add docstrings to these functions.

**Review by maskarb — COMMENTED** [review]
Just a couple things to add and then it's good to go.

File: koku/currency/__init__.py
Comment (ddonahue007) [reply koku/currency/__init__.py:1]: thx...

File: koku/api/settings/utils.py
Comment (ddonahue007) [reply koku/api/settings/utils.py:111]: Done

**Review by maskarb — APPROVED** [review]
lgtm!

File: koku/api/settings/settings.py
Comment (dlabrecq) [reply koku/api/settings/settings.py:286]: Just want to enure it doesn't make it to the browser console

**Pull request merged**
